### PR TITLE
Named Pipes: Support Message Read Mode

### LINKED ISSUE #5349: Named Pipes: Support Message Read Mode
**Context on Windows Named Pipes**

Windows has message-based and byte stream-based pipes, but there also is an additional concept of _read modes_. From [the docs](https://learn.microsoft.com/en-us/windows/win32/ipc/named-pipe-type-read-and-wait-modes):

> The type mode of a pipe determines how data is written to a named pipe. Data can be transmitted through a named pipe as either a stream of bytes or as a stream of messages.
> [...]
> The read mode of a pipe determines how data is read from a named pipe. The pipe server specifies the initial read mode for a pipe handle when calling [CreateNamedPipe](https://learn.microsoft.com/en-us/windows/desktop/api/Winbase/nf-winbase-createnamedpipea). Data can be read in byte-read mode or message-read mode. A handle to a byte-type pipe can be in byte-read mode only. A handle to a message-type pipe can be in either byte-read or message-read mode. For a message-type pipe, the read mode can be different for server and client handles to the same pipe instance.

**Is your feature request related to a problem? Please describe.**

Tokio allows the creation of message-based named pipes using [`.pipe_mode`](https://docs.rs/tokio/latest/tokio/net/windows/named_pipe/struct.ServerOptions.html#method.pipe_mode), but it currently always opens them in byte read mode. In other words, when calling `.read()`, a "random" slice of bytes is returned (which may not be aligned with message boundaries). This was unexpected to me, I expected to read messages from my message pipe.

The same problem applies to NamedPipeClient instances, which always operate in byte read mode at the moment.

**Describe the solution you'd like**
When calling `server_options.pipe_mode(PipeMode::Message)`, the pipe should not only be set to be a message type pipe, but it should also be opened in message read mode. 

For NamedPipeClient, an additional `.pipe_mode` method should be added to the builder to switch to message read mode (equivalent to the server).

**Describe alternatives you've considered**
The first design decision here is whether tokio should expose Windows' distinction between pipe type and read mode. In other words, tokio could alternatively expose two distinct `.pipe_mode` and `.read_mode` builder methods. I personally feel that this comes at the risk of introducting subtle bugs (people will forget to also set `.read_mode`, which isn't immediately obvious).

The second design decision is whether ClientNamedPipe should 1) automatically query the pipe type and use the equivalent read mode, or 2) have an explicit `.pipe_mode` builder method. The explicit version seems more reasonable to me, but I don't have strong opinions on this either.

**Additional context**
I will follow up with a PR for the proposed changes momentarily. :)

### PR DESCRIPTION
**Pull request by mhils**
This PR fixes #5349.

### CODE REVIEW

**Comment by mhils** [comment]
I've rebased this onto latest master, please let me know if there's anything else I can help with to get this shipped. 😃

File: tokio/src/net/windows/named_pipe.rs [hunk 1]
```diff
@@ -2588,13 +2619,13 @@ mod test {
...
-        assert_eq!(opts.pipe_mode, PIPE_TYPE_MESSAGE);
+        assert_eq!(opts.pipe_mode, PIPE_TYPE_MESSAGE | PIPE_READMODE_MESSAGE);
 
         opts.reject_remote_clients(true);
         opts.pipe_mode(PipeMode::Message);
         assert_eq!(
             opts.pipe_mode,
-            PIPE_TYPE_MESSAGE | PIPE_REJECT_REMOTE_CLIENTS
+            PIPE_TYPE_MESSAGE | PIPE_READMODE_MESSAGE | PIPE_REJECT_REMOTE_CLIENTS
         );
```
Comment (Darksonn) [inline tokio/src/net/windows/named_pipe.rs:2629]: What happens here if we change the pipe mode back to byte? Is `PIPE_READMODE_MESSAGE` still set?

Comment (mhils) [reply tokio/src/net/windows/named_pipe.rs:2629]: `PIPE_READMODE_MESSAGE` is cleared. I've pushed c582f04c to add that to the test. :)

I have another commit (658c3f85f710bb039a4b097cf2ed8808def92ddc) queued after this to refactor the code to not use bitfields (#5359).

**Review by Darksonn — APPROVED** [review]
Thanks.

**Pull request merged**